content/doc/book/using/long-running-agent-process.adoc

```diff
@@ -0,0 +1,43 @@
+---
+layout: section
+title: Long running agent process
+---
+
+= Long running agent process
+
+To reliably kill processes spawned by a job during a build, Jenkins
+contains a bit of native code to list up such processes and kill them.
+This is tested on several platforms and architectures, but if you find a show-stopper problem because of this, you can disable this feature by setting a Java property named "hudson.util.ProcessTree.disable" to the value "true".
+
+This can be done as a parameter to the "java" binary when starting Jenkins:
+
+[source,groovy]
+----
+ java -Dhudson.util.ProcessTree.disable=true -jar jenkins.war
+----
+
+Depending on how you run your container, this can be different. In the distributed build environment, this system property needs to be set on agent JVMs.
+
+[[ProcessTreeKiller-Howitworks]]
+== How it works
+
+The ProcessTreeKiller takes advantage of the fact that by default a new process gets a copy of the environment variables of its
+spawning/creating process.
+
+It sets a specific environment variable in the process executing the
+build job. Later, when the user requests to stop the build job's process it gets a list of all processes running on the computer and their environment variables and looks for the environment variable that it initially set for the build job's process.
+
+Every job with that environment variable in its environment is then
+terminated.
+
+[[ProcessTreeKiller-Ifyourbuildwantstoleaveadaemonrunningbehind...]]
+== If your build wants to leave a daemon running behind...
+
+A convenient way to achieve that is to change the environment variable `BUILD_ID` which Jenkins's ProcessTreeKiller is looking for. This will cause Jenkins to assume that your daemon is not spawned by the Jenkins build. For example:
+
+[source,groovy]
+----
+BUILD_ID=dontKillMe /usr/apache/bin/httpd
+----
+
+In case of Jenkins Pipeline use `+JENKINS_NODE_COOKIE+` instead of `+BUILD_ID+`
```

Comment:
See the comments in https://github.com/jenkinsci/workflow-durable-task-step-plugin/pull/39 for the details that I've tried to reduce to this phrasing.

I would love to have comments or corrections on my updated phrasing from @bitwiseman , @dwnusbaum , or @jglick.  I believe I've stated it correctly, but they may see mistakes in my corrections.

```suggestion
In case of Jenkins Pipeline use `+JENKINS_NODE_COOKIE+` instead of `+BUILD_ID+`.
Parallel stages of the Jenkins Pipeline may run on the same agent and will use the same `+BUILD_ID+` within a specific Pipeline run.
If `+BUILD_ID+` is set instead of `+JENKINS_NODE_COOKIE+`, then parallel Pipeline stages running on the same agent may stop processes from other stages of the same build.
Use `+JENKINS_NODE_COOKIE+` instead of `+BUILD_ID+` so that processes in parallel Pipeline stages are not unexpectedly stopped by the completion of other stages on that agent.
```

Reply:
I do not recall offhand and I would not expect any text about this from the wiki to be accurate. Would need to be experimentally revalidated.

Reply:
@LarrySul are you willing to do the experimental validation of my proposed text additions?

Reply:
@MarkEWaite yes I'm open to doing the experimental validations

Reply:
That's great @LarrySul.  Thanks

Things to test include:

* Is `JENKINS_NODE_COOKIE` defined inside a Pipeline stage? (should confirm that it is)
* Are processes on the same agent that are started from different stages of the same job stopped when a single stage stops?
* Are processes on the same agent that are started from different stages of the same job stopped when the pipeline has completed and all stages are stopped?
* If `JENKINS_NODE_COOKIE` is defined with a different value in a stage, do long running processes started in that stage continue running after the stage is complete?
* If `JENKINS_NODE_COOKIE` is defined with a different value in a stage, do long running processes started in that stage continue running after the job is complete?
* If `BUILD_ID` is defined with a different value in a stage but `JENKINS_NODE_COOKIE` is unmodified, are long running processes from a stage stopped when the stage completes?
* If `BUILD_ID` is defined with a different value in a stage but `JENKINS_NODE_COOKIE` is unmodified, are long running processes from a stage stopped when the job completes?